test/common/buffer/buffer_test.cc

```diff
@@ -14,6 +14,260 @@ namespace Envoy {
+    committed = other_slice->commit(other_reservation);
+    EXPECT_TRUE(committed);
+  }
+
+  {
+    // Try to reserve more space than is available in the slice.
+    uint64_t reservable_size = slice->reservableSize();
+    Slice::Reservation reservation = slice->reserve(reservable_size + 1);
+    expectReservationSuccess(reservation, *slice, reservable_size);
+    bool committed = slice->commit(reservation);
+    expectCommitSuccess(committed, *slice, initial_capacity, 0);
+  }
+
+  {
+    // Now that the view has no more reservable space, verify that it rejects
+    // subsequent reservation requests.
+    Slice::Reservation reservation = slice->reserve(1);
+    expectReservationFailure(reservation, *slice, 0);
+  }
+}
+
+TEST_F(OwnedSliceTest, Drain) {
+  // Create a slice and commit all the available space.
+  auto slice = OwnedSlice::create(100);
+  Slice::Reservation reservation = slice->reserve(slice->reservableSize());
+  bool committed = slice->commit(reservation);
+  EXPECT_TRUE(committed);
+  EXPECT_EQ(0, slice->reservableSize());
+
+  // Drain some data from the front of the view and verify that the data start moves accordingly.
+  const uint8_t* original_data = static_cast<const uint8_t*>(slice->data());
+  uint64_t original_size = slice->dataSize();
+  slice->drain(0);
+  EXPECT_EQ(original_data, slice->data());
+  EXPECT_EQ(original_size, slice->dataSize());
+  slice->drain(10);
+  EXPECT_EQ(original_data + 10, slice->data());
+  EXPECT_EQ(original_size - 10, slice->dataSize());
+  slice->drain(50);
+  EXPECT_EQ(original_data + 60, slice->data());
+  EXPECT_EQ(original_size - 60, slice->dataSize());
+
+  // Drain all the remaining data.
+  slice->drain(slice->dataSize());
+  EXPECT_EQ(0, slice->dataSize());
+  EXPECT_EQ(original_size, slice->reservableSize());
+}
+
+TEST(UnownedSliceTest, CreateDelete) {
+  constexpr char input[] = "hello world";
+  bool release_callback_called = false;
+  BufferFragmentImpl fragment(
+      input, sizeof(input) - 1,
+      [&release_callback_called](const void*, size_t, const BufferFragmentImpl*) {
+        release_callback_called = true;
+      });
+  auto slice = std::make_unique<UnownedSlice>(fragment);
+  EXPECT_EQ(11, slice->dataSize());
+  EXPECT_EQ(0, slice->reservableSize());
+  EXPECT_EQ(0, memcmp(slice->data(), input, slice->dataSize()));
+  EXPECT_FALSE(release_callback_called);
+  slice.reset(nullptr);
+  EXPECT_TRUE(release_callback_called);
+}
+
+TEST(SliceDequeTest, CreateDelete) {
+  bool slice1_deleted = false;
+  bool slice2_deleted = false;
+  bool slice3_deleted = false;
+
+  {
+    // Create an empty deque.
+    SliceDeque slices;
+    EXPECT_TRUE(slices.empty());
+    EXPECT_EQ(0, slices.size());
+
+    // Append a view to the deque.
+    const std::string slice1 = "slice1";
+    slices.emplace_back(
+        std::make_unique<DummySlice>(slice1, [&slice1_deleted]() { slice1_deleted = true; }));
+    EXPECT_FALSE(slices.empty());
+    EXPECT_EQ(1, slices.size());
```

Comment:
This particular one (and similar cases) should be an ASSERT_EQ because you want to exit this test-method if it fails, rather than crashing on the slices.front() below, thus allowing the other test-methods to still run.